cirq-core/cirq/qis/clifford_tableau.py

```diff
@@ -137,30 +138,47 @@ class CliffordTableau(StabilizerState):
     an eigenoperator of the state vector with eigenvalue one: P|psi> = |psi>.
     """
 
-    def __init__(self, num_qubits, initial_state: int = 0):
+    def __init__(
+        self,
+        num_qubits,
+        initial_state: int = 0,
+        rs: Optional[np.ndarray] = None,
+        xs: Optional[np.ndarray] = None,
+        zs: Optional[np.ndarray] = None,
+    ):
         """Initializes CliffordTableau
         Args:
             num_qubits: The number of qubits in the system.
             initial_state: The computational basis representation of the
                 state as a big endian int.
         """
         self.n = num_qubits
+        self.initial_state = initial_state
+
+        if rs is None:
+            # The last row (`2n+1`-th row) is the scratch row used in _measurement
+            # computation process only. It should not be exposed to external usage.
+            self._rs = np.zeros(2 * self.n + 1, dtype=bool)
+            for (i, val) in enumerate(
+                big_endian_int_to_digits(self.initial_state, digit_count=num_qubits, base=2)
+            ):
+                self._rs[self.n + i] = bool(val)
+        else:
+            self._rs = rs
```

Comment:
I think we should do two additional things. First, we should check that rs, xs and zs have the right shapes where the right shape is one without the scratchpad row. Second, we should add the scratchpad row.

(The reason I think we shouldn't expect the scratchpad row is that it is an internal implementation detail. I think it's ok to allow the scratchpad row to be provided by the caller as optimization.)

(Applies below to xs and zs, too.)

Reply:
Done.